compiler-core/src/build/package_compiler.rs

```diff
@@ -120,32 +135,39 @@ where
         )
         .run()?;
 
-        // Load the cached modules that have previously been compiled
-        for module in loaded.cached.into_iter() {
-            _ = existing_modules.insert(module.name.clone(), module.clone());
-        }
-
-        if !loaded.to_compile.is_empty() {
-            // Print that work is being done
-            if self.perform_codegen {
-                telemetry.compiling_package(&self.config.name);
-            } else {
-                telemetry.checking_package(&self.config.name)
+        let modules = match self.modules_compilation {
+            // If we're asked to ignore the project's modules we're going to
+            // perform the codegeneration step on an empty vector of modules.
+            ModulesCompilation::IgnoreModules => Ok(vec![]),
+            ModulesCompilation::CompileAll => {
+                // Load the cached modules that have previously been compiled
+                for module in loaded.cached.into_iter() {
+                    _ = existing_modules.insert(module.name.clone(), module.clone());
+                }
+
+                if !loaded.to_compile.is_empty() {
+                    // Print that work is being done
+                    if self.perform_codegen {
+                        telemetry.compiling_package(&self.config.name);
+                    } else {
+                        telemetry.checking_package(&self.config.name)
+                    }
+                }
+
+                // Type check the modules that are new or have changed
+                tracing::info!(count=%loaded.to_compile.len(), "analysing_modules");
+                analyse(
+                    &self.config,
+                    self.target.target(),
+                    self.mode,
+                    &self.ids,
+                    loaded.to_compile,
+                    existing_modules,
+                    warnings,
+                    self.target_support,
+                )
             }
-        }
-
-        // Type check the modules that are new or have changed
-        tracing::info!(count=%loaded.to_compile.len(), "analysing_modules");
-        let modules = analyse(
-            &self.config,
-            self.target.target(),
-            self.mode,
-            &self.ids,
-            loaded.to_compile,
-            existing_modules,
-            warnings,
-            self.target_support,
-        )?;
+        }?;
```

Comment:
Finding this a bit hard to follow as there's lots of indentation. How about instead we only call the package loader if we want to load modules, and use empty vectors otherwise.

Reply:
Ok! I've refactored it into a different function, now it should read better :)

Reply:
Can we not do the thing I asked for with the loader? I would much prefer to not load them at at all instead of loading and then dropping them. I think it would read much clearer too.